### PR TITLE
Applys explcit includes ruff file

### PR DESCRIPTION
**Pull request by CyclingNinja**
We've selected specific `.ruff.toml` rule categories to apply, this is the inital test of them

### CODE REVIEW

**Comment by CyclingNinja** [comment]
autofix

File: .ruff.toml [hunk 1]
```diff
@@ -12,20 +12,45 @@ select = [
...
+  "PERF",
+  "RUF",
 ]
 extend-ignore = [
   "E712",
   "E721",
-  # pycodestyle (E, W)
-  "E501", # LineTooLong                             # TODO! fix
+   # flake8-pie
+  "PIE808", # Disallow passing 0 as the first argument to range
```
Comment (nabobalis) [inline .ruff.toml:39]: What prompted this rule?

File: .ruff.toml [hunk 2]
```diff
@@ -38,6 +63,11 @@ extend-ignore = [
 "docs/*.py" = [
   "INP001",  # Implicit-namespace-package. The examples are not a package.
 ]
+"examples/*.py" = [
+  "T201",  # We need print in our examples
+  "E501",  # Line too long
```
Comment (nabobalis) [inline .ruff.toml:68]: We should ignore this forever.

File: ndcube/__init__.py [hunk 3]
```diff
@@ -14,4 +14,13 @@
 from .ndcube_sequence import NDCubeSequence, NDCubeSequenceBase
 from .version import version as __version__
 
-__all__ = ['NDCube', 'NDCubeSequence', "NDCollection", "ExtraCoords", "GlobalCoords", "ExtraCoordsABC", "GlobalCoordsABC", "NDCubeBase", "NDCubeSequenceBase", "__version__"]
+__all__ = ['NDCube',
```
Comment (nabobalis) [inline ndcube/__init__.py:17]: Was this auto formatted by something?

File: .ruff.toml [hunk 4]
```diff
@@ -12,20 +12,45 @@ select = [
   "E",
   "F",
   "W",
-  #"UP",
-  #"PT"
+  "N",
```
Comment (nabobalis) [inline .ruff.toml:15]: I would really be interested in maybe another PR that just enables all of the rules and we see what happens to ndcube.

File: .ruff.toml (hunk 1 above)
Comment (CyclingNinja) [reply .ruff.toml:39]: honestly don't know, it seems a weird thing to do so we're leaving it out

File: .ruff.toml [hunk 5]
```diff
@@ -12,20 +12,45 @@ select = [
...
-  # pycodestyle (E, W)
-  "E501", # LineTooLong                             # TODO! fix
+   # flake8-pie
+  "PIE808", # Disallow passing 0 as the first argument to range
+  # upgrades
+  "UP038", # Use | in isinstance - not compatible with models and is slower
+  # numpy
+  "NPY002", # TODO: migrate from np.random.rand to np.random.Generator
   # pytest (PT)
   "PT001", # Always use pytest.fixture()
```
[the remark this answers is not in the document]
Comment (CyclingNinja) [reply .ruff.toml:45]: I dunno, it might have to be a work in progress, PY011 has a shit load to unravel

File: .ruff.toml (hunk 4 above)
Comment (CyclingNinja) [reply .ruff.toml:15]: I can do this for you, there'll be a catastrophic amount of changes to be made

File: .ruff.toml (hunk 1 above)
Comment (nabobalis) [reply .ruff.toml:39]: To disallow 0 as a start input to range? Doesn't range default to 0?

File: .ruff.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (nabobalis) [reply .ruff.toml:45]: What does PY011 do?

File: .ruff.toml (hunk 2 above)
Comment (nabobalis) [reply .ruff.toml:71]: FOr everything.

File: ndcube/__init__.py (hunk 3 above)
Comment (nabobalis) [reply ndcube/__init__.py:17]: Ah!

File: .ruff.toml (hunk 4 above)
Comment (nabobalis) [reply .ruff.toml:15]: I don't mean for you to fix it. 

Just enable all the rules, autofix what can be fixed, open that PR and then do nothing else. Don't fix any fails yourself. I want to see what can be autofixed, what can't be and what needs to be fixed manually.

File: .ruff.toml (hunk 1 above)
Comment (CyclingNinja) [reply .ruff.toml:39]: Yeah exactly, that seems like a thing anyone who writes in Python would expect

File: .ruff.toml (hunk 2 above)
Comment (CyclingNinja) [reply .ruff.toml:71]: interesting, I'm not opposed, there's no good reason for the max line length anymore other than readability

File: .ruff.toml (hunk 4 above)
Comment (CyclingNinja) [reply .ruff.toml:15]: So I opened the said PR and ran the unsafe-fixes on it, have a lookse

File: .ruff.toml (hunk 2 above)
Comment (nabobalis) [reply .ruff.toml:71]: In an ideal world, the auto formatter deals with line length by itself. It isn't something we need the linter to worry about.

File: .ruff.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (CyclingNinja) [reply .ruff.toml:45]: PY011 is test Value Raises messages should be shorter and there are quite a few throught the testing base

File: .ruff.toml (hunk 2 above)
Comment (CyclingNinja) [reply .ruff.toml:71]: yeah reasonable, you'd have that in the pre-commit?

Comment (nabobalis) [reply .ruff.toml:71]: That would be the plan in the future.

File: .ruff.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Cadair) [reply .ruff.toml:45]: Adding things to ignores is what they are for, just because there's a linter rule for it doesn't make it "correct" there are some rules in ruff which directly conflict with other ruff rules.

These pytest rules are the ones I ignore because they aren't the style we use anywhere in sunpy so why on earth would we adopt it now?

File: .ruff.toml (hunk 2 above)
Comment (Cadair) [reply .ruff.toml:71]: I am also +1 in not having ruff lint line length, we should leave it to a formatter in the future.

File: .ruff.toml (hunk 1 above)
Comment (Cadair) [reply .ruff.toml:39]: > explicit is better than implicit

I don't know why you would have a linting rule which disallows being more explicit.

Comment (nabobalis) [reply .ruff.toml:39]: Its the same as setting None as the second arg to get when its the default. Why?

File: .ruff.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Cadair) [reply .ruff.toml]: So what? It's not the style we want to have?

File: .cruft.json [hunk 6]
```diff
@@ -16,6 +16,7 @@
       "enable_dynamic_dev_versions": "y",
       "include_example_code": "n",
       "include_cruft_update_github_workflow": "y",
+      "use_extended_ruff_linting": "n",
```
Comment (Cadair) [inline .cruft.json:19]: @CyclingNinja you need to do the variable update thingy here.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 repos:
     # This should be before any formatting hooks like isort
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: "v0.7.2"
+    rev: "v0.7.1"
```
Comment (nabobalis) [inline .pre-commit-config.yaml:4]:
```suggestion
    rev: "v0.7.2"
```

File: .ruff.toml [hunk 8]
```diff
@@ -12,57 +12,77 @@ select = [
...
 ]
-"__init__.py" = [
-  "E402", # Module level import not at top of cell
-  "F401", # Unused import
-  "F403", # from {name} import * used; unable to detect undefined names
-  "F405", # {name} may be undefined, or defined from star imports
-]
-"test_*.py" = [
-  "E402", # Module level import not at top of cell
-]
```
Comment (nabobalis) [inline .ruff.toml:65]: I think we want these versions instead of the flat ones below.

File: .ruff.toml [hunk 9]
```diff
@@ -12,57 +12,77 @@ select = [
...
+"examples/*.py" = [
+  "T201",  # We need print in our examples
+  "E501",  # Line too long
+  "INP001",  # implicit namespace package
 ]
-"examples/**.py" = [
+"ndcube/*.py" = [
   "T201", # allow use of print in examples
   "INP001", # File is part of an implicit namespace package.
 ]
```
Comment (nabobalis) [inline .ruff.toml:83]: Do we need this section?

File: ndcube/__init__.py [hunk 10]
```diff
@@ -14,4 +14,13 @@
...
+           'NDCubeSequence',
+           "NDCollection",
+           "ExtraCoords",
+           "GlobalCoords",
+           "ExtraCoordsABC",
+           "GlobalCoordsABC",
+           "NDCubeBase",
+           "NDCubeSequenceBase",
+           "__version__"]
```
Comment (nabobalis) [inline ndcube/__init__.py:26]:
```suggestion
__all__ = [
    'NDCube',
    'NDCubeSequence',
    "NDCollection",
    "ExtraCoords",
    "GlobalCoords",
    "ExtraCoordsABC",
    "GlobalCoordsABC",
    "NDCubeBase",
    "NDCubeSequenceBase",
    "__version__",
]
```

File: ndcube/_dev/scm_version.py [hunk 11]
```diff
@@ -1,5 +1,6 @@
 # Try to use setuptools_scm to get the current version; this is only used
 # in development installations from the git repository.
+
```
Comment (Cadair) [inline ndcube/_dev/scm_version.py:3]: This doesn't match the template.
```suggestion
```

File: ndcube/extra_coords/table_coord.py [hunk 12]
```diff
@@ -20,7 +20,11 @@
 except ImportError:
     pass
 
-__all__ = ['TimeTableCoordinate', 'SkyCoordTableCoordinate', 'QuantityTableCoordinate', "BaseTableCoordinate", "MultipleTableCoordinate"]
+__all__ = ['TimeTableCoordinate',
+           'SkyCoordTableCoordinate',
+           'QuantityTableCoordinate',
+           'BaseTableCoordinate',
+           'MultipleTableCoordinate']
```
Comment (nabobalis) [inline ndcube/extra_coords/table_coord.py:27]:
```suggestion
__all__ = [
    'TimeTableCoordinate',
    'SkyCoordTableCoordinate',
    'QuantityTableCoordinate',
    'BaseTableCoordinate',
    'MultipleTableCoordinate',
]
```

File: ndcube/extra_coords/table_coord.py [hunk 13]
```diff
@@ -136,7 +140,7 @@ def _generate_tabular(lookup_table, interpolation='linear', points_unit=u.pix, *
         raise TypeError("lookup_table must be a Quantity.")  # pragma: no cover
 
     ndim = lookup_table.ndim
-    TabularND = tabular_model(ndim, name=f"Tabular{ndim}D")
+    tabular_nd = tabular_model(ndim, name=f"Tabular{ndim}D")
```
Comment (Cadair) [inline ndcube/extra_coords/table_coord.py:143]: `tabular_model` returns a class object so this should be caps.
```suggestion
    TabularND = tabular_model(ndim, name=f"Tabular{ndim}D")
```

File: ndcube/extra_coords/table_coord.py [hunk 14]
```diff
@@ -151,7 +155,7 @@ def _generate_tabular(lookup_table, interpolation='linear', points_unit=u.pix, *
     if len(lookup_table) == 1:
         t = Length1Tabular(points, lookup_table, **kwargs)
     else:
-        t = TabularND(points, lookup_table, **kwargs)
+        t = tabular_nd(points, lookup_table, **kwargs)
```
Comment (Cadair) [inline ndcube/extra_coords/table_coord.py:158]:
```suggestion
        t = TabularND(points, lookup_table, **kwargs)
```

File: ndcube/extra_coords/table_coord.py [hunk 15]
```diff
@@ -921,11 +923,16 @@ def dropped_world_dimensions(self):
...
+        dropped_world_dimensions["world_axis_names"] += [name or None for name in
+                                                         dropped_multi_table.frame.axes_names]
+        dropped_world_dimensions["world_axis_physical_types"] += (
+            list(dropped_multi_table.frame.axis_physical_types))
+        dropped_world_dimensions["world_axis_units"] += [u.to_string() for u in
+                                                         dropped_multi_table.frame.unit]
+        dropped_world_dimensions["world_axis_object_components"] += \
+            dropped_multi_table.frame._world_axis_object_components
+        dropped_world_dimensions["world_axis_object_classes"].update(
+            dropped_multi_table.frame._world_axis_object_classes)
```
Comment (nabobalis) [inline ndcube/extra_coords/table_coord.py:936]: I would just restore this to all being one line.

File: ndcube/tests/helpers.py [hunk 16]
```diff
@@ -34,10 +34,17 @@ def get_hash_library_name():
     Generate the hash library name for this env.
     """
     ft2_version = f"{mpl.ft2font.__freetype_version__.replace('.', '')}"
-    animators_version = "dev" if (("dev" in mpl_animators.__version__) or ("rc" in mpl_animators.__version__)) else mpl_animators.__version__.replace('.', '')
-    mpl_version = "dev" if (("dev" in mpl.__version__) or ("rc" in mpl.__version__)) else mpl.__version__.replace('.', '')
-    astropy_version = "dev" if (("dev" in astropy.__version__) or ("rc" in astropy.__version__)) else astropy.__version__.replace('.', '')
-    return f"figure_hashes_mpl_{mpl_version}_ft_{ft2_version}_astropy_{astropy_version}_animators_{animators_version}.json"
```
Comment (nabobalis) [inline ndcube/tests/helpers.py:40]: Revert this to one line.

File: ndcube/wcs/wrappers/tests/test_resampled_wcs.py
Comment (Cadair) [inline ndcube/wcs/wrappers/tests/test_resampled_wcs.py:1]: I'd revert this whole file.

**Review by Cadair — CHANGES_REQUESTED** [review]

File: ndcube/visualization/mpl_sequence_plotter.py [hunk 17]
```diff
@@ -55,7 +54,10 @@ def animate(self, sequence_axis_coords=None, sequence_axis_unit=None, **kwargs):
             The unit in which the sequence_axis_coordinates should be displayed.
             If None, the default unit will be used.
         """
-        return SequenceAnimator(self._ndcube, sequence_axis_coords=sequence_axis_coords, sequence_axis_unit=sequence_axis_unit, **kwargs)
```
Comment (nabobalis) [inline ndcube/visualization/mpl_sequence_plotter.py:58]: I would make this one line again.

**Review by nabobalis — APPROVED** [review]

**Review by Cadair — APPROVED** [review]
